webrtc/README.md

```diff
@@ -0,0 +1,431 @@
+            - [Open Questions](#open-questions-1)
+    - [Multiplexing](#multiplexing)
+        - [Ordering](#ordering)
+    - [Connection Security](#connection-security)
+        - [Open Questions](#open-questions-2)
+    - [General Open Questions](#general-open-questions)
+    - [Previous, ongoing and related work](#previous-ongoing-and-related-work)
+- [FAQ](#faq)
+
+<!-- markdown-toc end -->
+
+## Motivation
+
+1. **No need for trusted TLS certificates.** Enable browsers to connect to
+   public server nodes without those server nodes providing a TLS certificate
+   within the browsers trustchain. Note that we can not do this today with our
+   Websocket transport as the browser requires the remote to have a trusted TLS
+   certificate. Nor can we establish a plain TCP or QUIC connection from within
+   a browser.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   server node to connect even though one or both are behind a NAT / firewall.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Addressing
+
+WebRTC multiaddresses are composed of an IP and UDP address component, followed
+by `/webrtc` and a multihash of the certificate that the node uses.
+
+Examples:
+- `/ip4/1.2.3.4/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+- `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+
+The TLS certificate fingerprint in `/certhash` is a
+[multibase](https://github.com/multiformats/multibase) encoded
+[multihash](https://github.com/multiformats/multihash).
+
+For compatibility implementations MUST support hash algorithm
+[`sha-256`](https://github.com/multiformats/multihash) and base encoding
+[`base64url`](https://github.com/multiformats/multibase). Implementations MAY
+support other hash algorithms and base encodings, but they may not be able to
+connect to all other nodes.
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ instantiates a `RTCPeerConnection`. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+   _A_ MAY reuse the same certificate across `RTCPeerConnection`s. Though one
+   should note that reusing the certificate can be used to identify _A_ across
+   connections by on-path observers given that WebRTC uses TLS 1.2. Thus this
+   specification RECOMMENDS that _A_, i.e. the browser, does not reuse the same
+   certificate across `RTCPeerConnection`.
```

Comment:
I updated this section to recommend against reusing the TLS certificate based on the suggestion by @marten-seemann in https://github.com/libp2p/specs/pull/412#issuecomment-1255472209:

> * Reusing the same browser-generated certificate has some subtle privacy risks. Since the WebRTC handshake uses DTLS 1.2 (the problem was fixed in 1.3), the `Certificate` messages are sent in the clear, meaning that an on-path observer can correlate connections from the same browser to different endpoints. This is a privacy risk for people taking steps to mask their IP information, e.g. by using a VPN. In any case, this spec should discuss this, and maybe even recommend using different certificates for connections to different IP addresses.

@ckousik any objections? Can you make sure this is the case in https://github.com/little-bear-labs/js-libp2p-webrtc/ ?

Reply:
Can confirm https://github.com/little-bear-labs/js-libp2p-webrtc/blob/ckousik/upgradable/src/transport.ts#L67

Reply:
Thanks @ckousik. Resolving here.